src/borg/key.py

```diff
@@ -154,13 +156,82 @@ class AESKeyBase(KeyBase):
 
     PAYLOAD_OVERHEAD = 1 + 32 + 8  # TYPE + HMAC + NONCE
 
+    def __init__(self, repository):
+        super(AESKeyBase, self).__init__(repository)
+        self.key_nonce_file = os.path.join(get_key_nonces_dir(), self.repository.id_str)
+
     def id_hash(self, data):
         """Return HMAC hash using the "id" HMAC key
         """
         return hmac_sha256(self.id_key, data)
 
+    def get_local_free_nonce(self):
+        if os.path.exists(self.key_nonce_file):
+            with open(self.key_nonce_file, 'r') as fd:
+                return bytes_to_long(unhexlify(fd.read()))
+        else:
+            return None
+
+    def commit_local_nonce_reservation(self, next_unreserved, start_nonce):
+        if self.get_local_free_nonce() != start_nonce:
+            raise Error("nonce space reservation with mismatched previous state")
+        with SaveFile(self.key_nonce_file, binary=False) as fd:
+            fd.write(bin_to_hex(long_to_bytes(next_unreserved)))
+
+    def get_repo_free_nonce(self):
+        try:
+            repo_free_nonce = self.repository.get_free_nonce()
+        except InvalidRPCMethod as error:
+            # old server version, suppress further calls
+            self.get_repo_free_nonce = lambda: None
+            self.save_repo_end_of_nonce_reservation = lambda next_unreserved, start_nonce: None
+            return None
+        if repo_free_nonce is not None:
+            repo_free_nonce = bytes_to_long(repo_free_nonce)
+        return repo_free_nonce
+
+    def commit_repo_nonce_reservation(self, next_unreserved, start_nonce):
+        if start_nonce is None:
+            self.repository.commit_nonce_reservation(long_to_bytes(next_unreserved), None)
+        else:
+            self.repository.commit_nonce_reservation(long_to_bytes(next_unreserved), long_to_bytes(start_nonce))
+
+    def ensure_reservation(self, nonce_space_needed):
+        if not self.end_of_nonce_reservation:
+            repo_free_nonce = self.get_repo_free_nonce()
+            local_free_nonce = self.get_local_free_nonce()
+            next_nonce = max(x for x in (repo_free_nonce, local_free_nonce, self.manifest_nonce) if x is not None)
+            reservation_end = next_nonce + nonce_space_needed + NONCE_SPACE_RESERVATION
+            self.commit_repo_nonce_reservation(reservation_end, repo_free_nonce)
+            self.commit_local_nonce_reservation(reservation_end, local_free_nonce)
+            self.end_of_nonce_reservation = reservation_end
+            if self.enc_cipher.iv != PREFIX + long_to_bytes(next_nonce):
+                self.enc_cipher.reset(self.enc_key, PREFIX + long_to_bytes(next_nonce))
+        else:
+            next_nonce = bytes_to_long(self.enc_cipher.iv[8:])
```

Comment:
hmm, here it is also reading back the iv.

i am asking because reading back the iv was a bit dirty in openssl 1.0 and does not work any more in 1.1.
also, it doesn't work for gcm, iirc.